make sure auto-mounted subpath mount source is already mounted  …

Dieken opened the pull request:
For example, we have two filesystems, one is embedded into another:

```
/a/test		# first filesystem with a directory "/a/test/b2"
/a/test/b2	# second filesystem, notice now "/a/test/b2" is a new directory on this filesystem
```

For subpath mount "/a/test/b2",  `openat("/a/test", "b2")` gets directory "b2" on the first
filesystem, then "mount -c" will use this wrong directory as source directory.

`openat("/a/test", "b2/")` forces traversing into the second filesystem to get the correct source
directory for "mount -c".

This fixes issue https://github.com/kubernetes/kubernetes/issues/110818#issuecomment-1175736550

<hr/>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

When a subpath points to an embedded child filesystem, kubelet wrongly mounts directory from the parent filesystem, then when a file access happens on the subpath mountpoint, the correct child filesystem is mounted automatically by linux kernel, this leads to two mounts on the same subpath,  kubelet will fail to destroy the pod because it wrongly thinks unmounting the subpath mount point failed.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/110818#issuecomment-1175736550

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

Dieken commented:
/assign @saad-ali

Dieken commented:
/cc @ jingxu97 @ jsafrane @ msau42
/assign @ jsafrane

Dieken commented:
/cc @jingxu97 @jsafrane @msau42
/assign @jsafrane

249043822 commented:
/ok-to-test

Dieken commented:
/retest

Dieken commented:
/retest

Seems the snapshot errors are due to CI environment.

jingxu97 commented on pkg/volume/util/subpath/subpath_linux.go:
```diff
@@ -587,6 +587,17 @@ func doSafeOpen(pathname string, base string) (int, error) {
 
+		// make sure to open directory on child filesystem such as bind mounted directory
+		if fileFmt == syscall.S_IFDIR {
+			if err = syscall.Close(childFD); err != nil {
```
what is the purpose of this call?

Dieken replied:
Because this child directory will be opened again below, so childFD is closed here to avoid file descriptor leak.

jingxu97 commented on pkg/volume/util/subpath/subpath_linux.go:
```diff
@@ -587,6 +587,17 @@ func doSafeOpen(pathname string, base string) (int, error) {
+				return -1, fmt.Errorf("closing dir fd for %q failed: %v", currentPath, err)
+			}
+
+			if childFD, err = syscall.Openat(parentFD, seg+"/", openFDFlags|unix.O_CLOEXEC, 0); err != nil {
```
I don't know the details, but just comparing with the previous function, the only difference is seg vs seg+"/"? 
Is it ok we always use seg+"/"? 

```
childFD, err = syscall.Openat(parentFD, seg, openFDFlags|unix.O_CLOEXEC, 0)
```

Dieken replied:
As the commit log and the corresponding issue comment said,  if "child-dir" is a mount point for another file system, `openat("parent-dir", "child-dir")` actually open directory on file system "parent-dir", not the expected directory on file system "parent-dir/child-dir".

I got three versions of this patch:

(1) always open `seg+"/"`.

This is wrong, it breaks plain file bind mount.

(2) first try `seg+"/"`, if gets ENOTDIR error, try `seg` again.

This leads to security vulnerability and breaks Kubernetes CI tests, when `seg` is a symlink, `seg+"/"` will follow it, this behaviour conflicts with `O_NOFOLLOW` in `openFDFlags`.

(3) the current version,  first try `seg`, if it's a directory, try `seg+"/"` again to force traversing into a possible mount point.

Dieken committed:
Make sure auto-mounted subpath mount source is already mounted

For example, we have two filesystems, one is embedded into another:

/a/test		# first filesystem with a directory "/a/test/b2"
/a/test/b2	# not auto mounted yet second filesystem, notice "/a/test/b2" is
	        # a new directory on this filesystem after this filesystem is mounted

For subpath mount "/a/test/b2",  `openat("/a/test", "b2")` gets directory "b2" on the first
filesystem, then "mount -c" will use this wrong directory as source directory.

`fstatat("/a/test", "b2/")` forces triggering auto mount of second filesystem, so
`openat("/a/test", "b2")` gets correct source directory for "mount -c".

This fixes issue https://github.com/kubernetes/kubernetes/issues/110818#issuecomment-1175736550

References:

1. https://man7.org/linux/man-pages/man2/openat.2.html

   If pathname refers to an automount point that has not yet
   been triggered, so no other filesystem is mounted on it,
   then the call returns a file descriptor referring to the
   automount directory without triggering a mount.

2. https://man7.org/linux/man-pages/man2/open_by_handle_at.2.html

   name_to_handle_at() does not trigger a mount when the final
   component of the pathname is an automount point.  When a
   filesystem supports both file handles and automount points, a
   name_to_handle_at() call on an automount point will return with
   error EOVERFLOW without having increased handle_bytes.  This can
   happen since Linux 4.13 with NFS when accessing a directory which
   is on a separate filesystem on the server.  In this case, the
   automount can be triggered by adding a "/" to the end of the
   pathname.

Dieken commented:
@jingxu97  please review again, I pushed the fourth version,  use `fstatat("parent-dir", "child-dir/")` to force auto mount of "child-dir", this version resolves the race condition in the third version,  even the attack unmounts "child-dir/" before `openat("parent-dir", "child-dir")`, he can't fool kubelet to use a symlink "child-dir".

Dieken commented:
/retest

jingxu97 commented:
/assign @jsafrane @msau42

Dieken commented:
The snapshot test error seems not related to this PR.

The bug can be easily reproduced with steps at https://github.com/kubernetes/kubernetes/issues/110818#issuecomment-1175736550, the patch itself is obvious.

Could you review/merge this little patch and backport it to v1.20? Currently this issue affects k8s >= v1.19.15, v1.20.11, v1.21.10, v1.22.1, v1.23.0 by searching `git grep no-canonicalize $TAG:pkg/volume/util/subpath/subpath_linux.go`

@249043822 @jingxu97 @msau42 @jsafrane @saad-ali

249043822 commented:
Since the pr: https://github.com/kubernetes/kubernetes/pull/110980 has merged, now the pull-kubernetes-e2e-gce-storage-snapshot job should normal
/test pull-kubernetes-e2e-gce-storage-snapshot

jsafrane commented on pkg/volume/util/subpath/subpath_linux.go:
```diff
@@ -566,18 +566,22 @@ func doSafeOpen(pathname string, base string) (int, error) {
 			return -1, fmt.Errorf("path %s is outside of allowed base %s", currentPath, base)
 		}
 
+		// trigger auto mount, ignore error if not a directory
```
Please add a comment that `unix.Fstat("path/a")` below is not enough and extra `path/a/` here is needed to trigger automount in case `a` is an automounted dir.

Dieken replied:
Updated the comment and rebased on latest master branch, please check again, thanks!

Dieken commented:
got timeout :-(

/retest

jsafrane commented:
/lgtm
/approve

rajnikhil17 commented:
How this fix will be released?

Will it be a patch or are you going to port it on 1.25?

jsafrane commented:
It will be part of 1.25

jsafrane commented:
/retest

Dieken commented:
> How this fix will be released?
> 
> Will it be a patch or are you going to port it on 1.25?

Currently this issue affects k8s >= v1.19.15, v1.20.11, v1.21.10, v1.22.1, v1.23.0 by searching `git grep no-canonicalize $TAG:pkg/volume/util/subpath/subpath_linux.go`,  it would be appreciated if backport to these releases.

@jsafrane  @rajnikhil17

rajnikhil17 commented:
@Dieken We are in the 1.23.6 release
Will we get this Patch to 1.23.6?

@jsafrane Confirmed that it would be ported under the 1.25 release

andyzhangx commented:
we also have a user hitting this issue, I am now cherry-picking this PR to v1.22, 1.23, 1.24

rajnikhil17 commented:
@andyzhangx  What would be the version in 1.23 this PR will be cherry-picked.

andyzhangx commented:
> I am now cherry-picking this PR to v1.22, 1.23, 1.24

@rajnikhil17 it should be `1.23.12` if the cherry-pick PR could be merged in the next few days.

rajnikhil17 commented:
@andyzhangx  Is this fix cherry picked?
